File: cpp/mujoco/play_r1_isaac.cpp
// play_r1_isaac.cpp
// ---------------------------------------------------------------------------
// Puerto DIRECTO a C++ de play_r1_isaac.py: misma logica exacta (obs 405,
// history de 5 pasos, reordenamiento Isaac<->MuJoCo, PD local con
// TORQUE_LIMITS_MJ), pero usando la API C de MuJoCo + GLFW en vez de
// mujoco.viewer de Python. NO usa DDS ni unitree_sdk2 -- es standalone.
//
// Controles (identicos al script de Python):
//   flechas  : vx (arriba/abajo), wz/giro (izq/der)
//   Q / E    : vy lateral
//   ESPACIO  : detener (comando = 0)
//   R        : reset a la pose inicial
//   ESC      : salir
//   mouse    : rotar/paneo/zoom de camara (arrastrar + rueda)
//
// Uso:
//   ./play_r1_isaac <ruta_r1.xml> <ruta_r1_policy_v2.pt>
// ---------------------------------------------------------------------------

#include <mujoco/mujoco.h>
#include <GLFW/glfw3.h>
#include <torch/script.h>

#include <array>
#include <deque>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <chrono>
#include <thread>
#include <algorithm>
#include <string>

// ===========================================================================
// CONSTANTES  (copiadas 1:1 de play_r1_isaac.py)
// ===========================================================================
static constexpr int NUM_DOFS    = 24;
static constexpr int HISTORY_LEN = 5;
static constexpr int OBS_DIM     = 405;

static constexpr double SIM_DT      = 0.002;  // paso de fisica
static constexpr int    DECIMATION  = 10;     // 0.002*10 = 0.02s -> 50Hz de politica
static constexpr float  ACTION_SCALE    = 0.25f;
static constexpr float  SCALE_ANG_VEL   = 0.2f;
static constexpr float  SCALE_JOINT_VEL = 0.05f;

// accion_mujoco = accion_isaac[MUJOCO_FROM_ISAAC]  (el "array de oro")
static const std::array<int, NUM_DOFS> MUJOCO_FROM_ISAAC = {
    0, 3, 6, 10, 14, 18,  1, 4, 7, 11, 15, 19,
    2, 5, 8, 12, 16, 20, 22,  9, 13, 17, 21, 23
};
static std::array<int, NUM_DOFS> ISAAC_FROM_MUJOCO{}; // se calcula en main() (argsort)

// Pose por defecto en ORDEN MUJOCO
static const std::array<float, NUM_DOFS> DEFAULT_MJ = {
    -0.06f, 0.0f, 0.0f, 0.05f, -0.04f, 0.0f,
    -0.06f, 0.0f, 0.0f, 0.05f, -0.04f, 0.0f,
     0.0f,  0.0f,
     0.18f, 0.18f, 0.0f, 1.5f, 0.0f,
     0.18f,-0.18f, 0.0f, 1.5f, 0.0f
};
static std::array<float, NUM_DOFS> DEFAULT_ISAAC{}; // se calcula en main()

static const std::array<float, NUM_DOFS> STIFFNESS_MJ = {
    100,100,100,200,80,80,  100,100,100,200,80,80,
    250,250,  50,50,40,30,20,  50,50,40,30,20
};
static const std::array<float, NUM_DOFS> DAMPING_MJ = {
    5,5,5,8,5,5,  5,5,5,8,5,5,
    25,25,  5,5,4,4,4,  5,5,4,4,4
};
static const std::array<float, NUM_DOFS> TORQUE_LIMITS_MJ = {
    88,139,88,139,50,50,  88,139,88,139,50,50,
    88,50,  25,25,25,25,25,  25,25,25,25,25
};

// ===========================================================================
// UTILIDADES
// ===========================================================================
static void mjToIsaac(const std::array<float,NUM_DOFS>& mj, std::array<float,NUM_DOFS>& isaac) {
    for (int i = 0; i < NUM_DOFS; ++i) isaac[i] = mj[ISAAC_FROM_MUJOCO[i]];
}
static void isaacToMj(const std::array<float,NUM_DOFS>& isaac, std::array<float,NUM_DOFS>& mj) {
    for (int i = 0; i < NUM_DOFS; ++i) mj[i] = isaac[MUJOCO_FROM_ISAAC[i]];
}

// Equivalente a quat_rotate_inverse() de Python: rota v del frame mundo al
// frame del cuerpo. q en orden (w,x,y,z).
static void quatRotateInverse(const double q[4], const float v[3], float out[3]) {
    const float w = (float)q[0];
    const float qv[3] = { (float)q[1], (float)q[2], (float)q[3] };
    const float dot = qv[0]*v[0] + qv[1]*v[1] + qv[2]*v[2];
    const float cross[3] = {
        qv[1]*v[2] - qv[2]*v[1],
        qv[2]*v[0] - qv[0]*v[2],
        qv[0]*v[1] - qv[1]*v[0]
    };
    for (int i = 0; i < 3; ++i)
        out[i] = v[i]*(2.0f*w*w - 1.0f) - cross[i]*(2.0f*w) + qv[i]*(2.0f*dot);
}

// ===========================================================================
// ESTADO GLOBAL  (patron estandar de los ejemplos de MuJoCo: simple.cc/basic.c)
// ===========================================================================
static mjModel* m = nullptr;
static mjData*  d = nullptr;

static mjvCamera cam;
static mjvOption opt;
static mjvScene  scn;
static mjrContext con;

// mouse
static bool button_left = false, button_middle = false, button_right = false;
static double lastx = 0, lasty = 0;

// politica
static torch::jit::script::Module policy;

// comando [vx, vy, wz]
static std::array<float, 3> command = {0.f, 0.f, 0.f};

// last_action en ORDEN ISAAC (salida cruda de la red, paso anterior)
static std::array<float, NUM_DOFS> last_action_isaac{};

// history (deques de snapshots ya escalados, orden Isaac)
static std::deque<std::array<float,3>>        h_ang_vel, h_grav, h_cmd;
static std::deque<std::array<float,NUM_DOFS>> h_jpos, h_jvel, h_act;

static bool has_gyro = false;
static int  gyro_adr = -1;

// ===========================================================================
// RESET  (equivalente a R1IsaacPolicyEnv.reset())
// ===========================================================================
static void reset_sim() {
    mj_resetData(m, d);
    d->qpos[0] = 0.0; d->qpos[1] = 0.0; d->qpos[2] = 0.74;
    d->qpos[3] = 1.0; d->qpos[4] = 0.0; d->qpos[5] = 0.0; d->qpos[6] = 0.0;
    for (int i = 0; i < NUM_DOFS; ++i) d->qpos[7 + i] = DEFAULT_MJ[i];
    for (int i = 0; i < m->nv; ++i) d->qvel[i] = 0.0;
    mj_forward(m, d);

    last_action_isaac.fill(0.0f);
    h_ang_vel.assign(HISTORY_LEN, std::array<float,3>{});
    h_grav.assign(HISTORY_LEN, std::array<float,3>{});
    h_cmd.assign(HISTORY_LEN, std::array<float,3>{});
    h_jpos.assign(HISTORY_LEN, std::array<float,NUM_DOFS>{});
    h_jvel.assign(HISTORY_LEN, std::array<float,NUM_DOFS>{});
    h_act.assign(HISTORY_LEN, std::array<float,NUM_DOFS>{});
}

// ===========================================================================
// OBSERVACION  (equivalente a _build_obs())
// ===========================================================================
static void buildObs(float* obs /* [OBS_DIM] */) {
    std::array<float, NUM_DOFS> qpos_mj, qvel_mj;
    for (int i = 0; i < NUM_DOFS; ++i) {
        qpos_mj[i] = (float)d->qpos[7 + i];
        qvel_mj[i] = (float)d->qvel[6 + i];
    }
    const double* quat = &d->qpos[3]; // (w,x,y,z)

    std::array<float,3> ang_vel{};
    if (has_gyro) {
        ang_vel = { (float)d->sensordata[gyro_adr + 0],
                    (float)d->sensordata[gyro_adr + 1],
                    (float)d->sensordata[gyro_adr + 2] };
    } else {
        ang_vel = { (float)d->qvel[3], (float)d->qvel[4], (float)d->qvel[5] };
    }

    std::array<float,3> ang_vel_obs, grav_obs;
    for (int i = 0; i < 3; ++i) ang_vel_obs[i] = ang_vel[i] * SCALE_ANG_VEL;
    const float g_world[3] = {0.f, 0.f, -1.f};
    quatRotateInverse(quat, g_world, grav_obs.data());

    std::array<float,3> cmd_obs = command;

    std::array<float,NUM_DOFS> jpos_isaac, jvel_isaac, tmp;
    mjToIsaac(qpos_mj, tmp);
    for (int i = 0; i < NUM_DOFS; ++i) jpos_isaac[i] = tmp[i] - DEFAULT_ISAAC[i];
    mjToIsaac(qvel_mj, tmp);
    for (int i = 0; i < NUM_DOFS; ++i) jvel_isaac[i] = tmp[i] * SCALE_JOINT_VEL;

    h_ang_vel.push_back(ang_vel_obs); h_ang_vel.pop_front();
    h_grav.push_back(grav_obs);       h_grav.pop_front();
    h_cmd.push_back(cmd_obs);         h_cmd.pop_front();
    h_jpos.push_back(jpos_isaac);     h_jpos.pop_front();
    h_jvel.push_back(jvel_isaac);     h_jvel.pop_front();
    h_act.push_back(last_action_isaac); h_act.pop_front();

    int k = 0;
    for (auto& s : h_ang_vel) for (float x : s) obs[k++] = x; // 15
    for (auto& s : h_grav)    for (float x : s) obs[k++] = x; // 15
    for (auto& s : h_cmd)     for (float x : s) obs[k++] = x; // 15
    for (auto& s : h_jpos)    for (float x : s) obs[k++] = x; // 120
    for (auto& s : h_jvel)    for (float x : s) obs[k++] = x; // 120
    for (auto& s : h_act)     for (float x : s) obs[k++] = x; // 120
    // k == 405
}

// ===========================================================================
// PASO DE POLITICA  (equivalente a _policy_step())
// ===========================================================================
static std::array<float, NUM_DOFS> policyStep() {
    float obs[OBS_DIM];
    buildObs(obs);

    torch::Tensor obs_t = torch::from_blob(obs, {1, OBS_DIM}, torch::kFloat32).clone();
    torch::NoGradGuard ng;
    torch::Tensor out = policy.forward({obs_t}).toTensor().to(torch::kCPU).contiguous();
    const float* a = out.data_ptr<float>();

    std::array<float, NUM_DOFS> action_isaac, target_isaac, target_mj;
    for (int i = 0; i < NUM_DOFS; ++i) action_isaac[i] = a[i];
    last_action_isaac = action_isaac;

    for (int i = 0; i < NUM_DOFS; ++i)
        target_isaac[i] = DEFAULT_ISAAC[i] + ACTION_SCALE * action_isaac[i];
    isaacToMj(target_isaac, target_mj);
    return target_mj;
}

// ===========================================================================
// PD LOCAL con clamp de torque  (equivalente a _apply_pd())
// ===========================================================================
static void applyPd(const std::array<float, NUM_DOFS>& target_mj) {
    for (int i = 0; i < NUM_DOFS; ++i) {
        float q  = (float)d->qpos[7 + i];
        float qd = (float)d->qvel[6 + i];
        float torque = STIFFNESS_MJ[i] * (target_mj[i] - q) - DAMPING_MJ[i] * qd;
        torque = std::clamp(torque, -TORQUE_LIMITS_MJ[i], TORQUE_LIMITS_MJ[i]);
        d->ctrl[i] = torque;
    }
}

// ===========================================================================
// CALLBACKS DE GLFW  (teclado identico al key_cb de Python + mouse estandar
// de los ejemplos oficiales de MuJoCo para rotar/paneo/zoom de camara)
// ===========================================================================
static void keyboard(GLFWwindow* window, int key, int scancode, int act, int mods) {
    if (act != GLFW_PRESS) return; // solo en el flanco de bajada, como Python

    switch (key) {
        case GLFW_KEY_ESCAPE: glfwSetWindowShouldClose(window, GLFW_TRUE); return;
        case GLFW_KEY_UP:     command[0] += 0.1f; break;
        case GLFW_KEY_DOWN:   command[0] -= 0.1f; break;
        case GLFW_KEY_LEFT:   command[2] += 0.1f; break;
        case GLFW_KEY_RIGHT:  command[2] -= 0.1f; break;
        case GLFW_KEY_Q:      command[1] += 0.1f; break;
        case GLFW_KEY_E:      command[1] -= 0.1f; break;
        case GLFW_KEY_SPACE:  command.fill(0.0f); break;
        case GLFW_KEY_R:      reset_sim(); return;
        default: return;
    }
    for (auto& c : command) c = std::clamp(c, -1.0f, 1.0f);
    printf("\rComando  vx=%+.2f  vy=%+.2f  wz=%+.2f   ", command[0], command[1], command[2]);
    fflush(stdout);
}

static void mouse_button(GLFWwindow* window, int button, int act, int mods) {
    button_left   = glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT)   == GLFW_PRESS;
    button_middle = glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_MIDDLE) == GLFW_PRESS;
    button_right  = glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_RIGHT)  == GLFW_PRESS;
    glfwGetCursorPos(window, &lastx, &lasty);
}

static void mouse_move(GLFWwindow* window, double xpos, double ypos) {
    if (!button_left && !button_middle && !button_right) return;
    double dx = xpos - lastx, dy = ypos - lasty;
    lastx = xpos; lasty = ypos;

    int width, height;
    glfwGetWindowSize(window, &width, &height);
    bool mod_shift = (glfwGetKey(window, GLFW_KEY_LEFT_SHIFT) == GLFW_PRESS) ||
                      (glfwGetKey(window, GLFW_KEY_RIGHT_SHIFT) == GLFW_PRESS);

    mjtMouse action;
    if (button_right)
        action = mod_shift ? mjMOUSE_MOVE_H : mjMOUSE_MOVE_V;
    else if (button_left)
        action = mod_shift ? mjMOUSE_ROTATE_H : mjMOUSE_ROTATE_V;
    else
        action = mjMOUSE_ZOOM;

    mjv_moveCamera(m, action, dx / height, dy / height, &scn, &cam);
}

static void scroll(GLFWwindow* window, double xoffset, double yoffset) {
    mjv_moveCamera(m, mjMOUSE_ZOOM, 0, -0.05 * yoffset, &scn, &cam);
}

// ===========================================================================
// MAIN
// ===========================================================================
int main(int argc, char** argv) {
    std::string xml_path    = (argc > 1) ? argv[1] : "assets/r1.xml";
    std::string policy_path = (argc > 2) ? argv[2] : "policies/r1_policy_v2.pt";

    // inverso del array de oro
    for (int isaac = 0; isaac < NUM_DOFS; ++isaac)
        ISAAC_FROM_MUJOCO[MUJOCO_FROM_ISAAC[isaac]] = isaac;
    mjToIsaac(DEFAULT_MJ, DEFAULT_ISAAC);

    // --- cargar modelo ---
    char error[1000] = "";
    printf("Cargando modelo: %s\n", xml_path.c_str());
    m = mj_loadXML(xml_path.c_str(), nullptr, error, sizeof(error));
    if (!m) {
        fprintf(stderr, "ERROR cargando XML: %s\n", error);
        return 1;
    }
    m->opt.timestep = SIM_DT;
    d = mj_makeData(m);

    int gyro_id = mj_name2id(m, mjOBJ_SENSOR, "imu_ang_vel");
    if (gyro_id >= 0) {
        has_gyro = true;
        gyro_adr = m->sensor_adr[gyro_id];
    }

    // --- cargar politica ---
    printf("Cargando politica: %s\n", policy_path.c_str());
    policy = torch::jit::load(policy_path);
    policy.eval();

    reset_sim();

    // --- GLFW + visor MuJoCo (patron estandar simple.cc/basic.c) ---
    if (!glfwInit()) { fprintf(stderr, "No se pudo iniciar GLFW\n"); return 1; }
    GLFWwindow* window = glfwCreateWindow(1200, 900, "R1 - politica nativa Isaac (C++)", nullptr, nullptr);
    if (!window) { fprintf(stderr, "No se pudo crear la ventana\n"); glfwTerminate(); return 1; }
    glfwMakeContextCurrent(window);
    glfwSwapInterval(1);

    mjv_defaultCamera(&cam);
    mjv_defaultOption(&opt);
    mjv_defaultScene(&scn);
    mjr_defaultContext(&con);
    mjv_makeScene(m, &scn, 2000);
    mjr_makeContext(m, &con, mjFONTSCALE_150);

    cam.distance  = 3.0;
    cam.elevation = -20;
    cam.azimuth   = 180;

    glfwSetKeyCallback(window, keyboard);
    glfwSetMouseButtonCallback(window, mouse_button);
    glfwSetCursorPosCallback(window, mouse_move);
    glfwSetScrollCallback(window, scroll);

    printf("\n== R1 con politica NATIVA de Isaac (C++/MuJoCo directo) ==\n");
    printf("Flechas: vx/giro | Q/E: lateral | ESPACIO: parar | R: reset | ESC: salir\n\n");

    std::array<float, NUM_DOFS> target_mj;
    for (int i = 0; i < NUM_DOFS; ++i) target_mj[i] = DEFAULT_MJ[i];
    long step = 0;

    while (!glfwWindowShouldClose(window)) {
        auto t0 = std::chrono::steady_clock::now();

        if (step % DECIMATION == 0) target_mj = policyStep();
        applyPd(target_mj);
        mj_step(m, d);

        cam.lookat[0] = d->qpos[0];
        cam.lookat[1] = d->qpos[1];
        cam.lookat[2] = d->qpos[2];

        int width, height;
        glfwGetFramebufferSize(window, &width, &height);
        mjrRect viewport = {0, 0, width, height};
        mjv_updateScene(m, d, &opt, nullptr, &cam, mjCAT_ALL, &scn);
        mjr_render(viewport, &scn, &con);
        glfwSwapBuffers(window);
        glfwPollEvents();

        ++step;
        auto elapsed = std::chrono::steady_clock::now() - t0;
        auto target_dur = std::chrono::duration<double>(SIM_DT);
        auto sleep_dur = target_dur - elapsed;
        if (sleep_dur > std::chrono::duration<double>(0))
            std::this_thread::sleep_for(sleep_dur);
    }

    printf("\nFin\n");
    mjv_freeScene(&scn);
    mjr_freeContext(&con);
    mj_deleteData(d);
    mj_deleteModel(m);
    glfwTerminate();
    return 0;
}

File: scripts/r1/play_r1_isaac.py
#!/usr/bin/env python3
"""
play_r1_isaac.py — Corre la politica NATIVA del R1 (entrenada en Isaac Lab)e
dentro de MuJoCo, reemplazando el parche del adapter de G1.

Reconstruye exactamente la observacion que Isaac le daba a la politica:
  obs (405) = [ base_ang_vel(15), projected_gravity(15), velocity_commands(15),
                joint_pos_rel(120), joint_vel_rel(120), last_action(120) ]
  - Todo en ORDEN DE JOINTS DE ISAAC.
  - Cada termino con history de 5 pasos, aplanado [t-4, t-3, t-2, t-1, t].
  - Escalas: base_ang_vel x0.2, joint_vel_rel x0.05, el resto x1.0.

La salida de la politica (24 acciones, orden Isaac) se convierte a target con
  target_isaac = default_isaac + 0.25 * accion_isaac
luego se REORDENA a MuJoCo con el "array de oro" y se aplica como PD.

Uso:
  python play_r1_isaac.py
Teclas:
  flechas  : vx (arriba/abajo), wz/giro (izq/der)
  Q / E    : vy lateral
  ESPACIO  : detener (comando = 0)
  R        : reset a la pose inicial
  ESC      : salir
"""

# --- hace importable el paquete `amo` sin instalar nada ---
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
# ----------------------------------------------------------

from amo.math_utils import quat_rotate_inverse
from amo.paths import policy, scene

import os
os.environ.setdefault("MUJOCO_GL", "glfw")

import time
from collections import deque

import numpy as np
import mujoco
import mujoco.viewer
import torch

# =============================================================================
# RUTAS
# =============================================================================
XML_PATH    = scene("r1")
POLICY_PATH = policy("r1_v2")

# =============================================================================
# MAPEO DE JOINTS  (el "array de oro" verificado con check_r1.py)
# =============================================================================
# accion_mujoco = accion_isaac[MUJOCO_FROM_ISAAC]
MUJOCO_FROM_ISAAC = np.array(
    [0, 3, 6, 10, 14, 18, 1, 4, 7, 11, 15, 19, 2, 5, 8, 12, 16, 20, 22, 9, 13, 17, 21, 23]
)
# Inverso: vector_isaac = vector_mujoco[ISAAC_FROM_MUJOCO]
ISAAC_FROM_MUJOCO = np.argsort(MUJOCO_FROM_ISAAC)

NUM_DOFS = 24

# =============================================================================
# POSE POR DEFECTO  (en ORDEN MUJOCO, igual que el init_state del CFG)
# =============================================================================
# left leg(6), right leg(6), waist(2), left arm(5), right arm(5)
DEFAULT_MJ = np.array([
    -0.06, 0.0, 0.0, 0.05, -0.04, 0.0,   # pierna izq
    -0.06, 0.0, 0.0, 0.05, -0.04, 0.0,   # pierna der
     0.0, 0.0,                           # cintura (roll, yaw)
     0.18, 0.18, 0.0, 1.5, 0.0,          # brazo izq
     0.18,-0.18, 0.0, 1.5, 0.0,          # brazo der
], dtype=np.float32)
DEFAULT_ISAAC = DEFAULT_MJ[ISAAC_FROM_MUJOCO]

# =============================================================================
# GANANCIAS PD  (en ORDEN MUJOCO, iguales a los actuadores del CFG de Isaac)
# =============================================================================
STIFFNESS_MJ = np.array([
    100, 100, 100, 200, 80, 80,
    100, 100, 100, 200, 80, 80,
    250, 250,
    50, 50, 40, 30, 20,
    50, 50, 40, 30, 20,
], dtype=np.float32)

DAMPING_MJ = np.array([
    5, 5, 5, 8, 5, 5,
    5, 5, 5, 8, 5, 5,
    25, 25,
    5, 5, 4, 4, 4,
    5, 5, 4, 4, 4,
], dtype=np.float32)

TORQUE_LIMITS_MJ = np.array([
    88, 139, 88, 139, 50, 50,
    88, 139, 88, 139, 50, 50,
    88, 50,
    25, 25, 25, 25, 25,
    25, 25, 25, 25, 25,
], dtype=np.float32)

# =============================================================================
# PARAMETROS DE CONTROL  (para casar con Isaac: control a 50 Hz)
# =============================================================================
SIM_DT      = 0.002            # paso de fisica de MuJoCo
DECIMATION  = 10               # 0.002 * 10 = 0.02 s  ->  50 Hz de politica
ACTION_SCALE = 0.25            # del JointPositionActionCfg
HISTORY_LEN = 5

# Escalas de observacion (Isaac)
SCALE_ANG_VEL  = 0.2
SCALE_JOINT_VEL = 0.05


# =============================================================================
# ENTORNO
# =============================================================================
class R1IsaacPolicyEnv:
    def __init__(self, xml_path=XML_PATH, policy_path=POLICY_PATH, device="cpu"):
        self.device = device

        print(f"Cargando modelo: {xml_path}")
        self.model = mujoco.MjModel.from_xml_path(xml_path)
        self.model.opt.timestep = SIM_DT
        self.data = mujoco.MjData(self.model)

        print(f"Cargando politica: {policy_path}")
        self.policy = torch.jit.load(policy_path, map_location=device)
        self.policy.eval()

        # Comando [vx, vy, wz]
        self.command = np.zeros(3, dtype=np.float32)

        # last_action en ORDEN ISAAC (salida cruda de la red, paso anterior)
        self.last_action_isaac = np.zeros(NUM_DOFS, dtype=np.float32)

        # Buffers de history (deque por termino, lleno de ceros al inicio)
        # Cada entrada es el snapshot YA ESCALADO de ese paso.
        self.h_ang_vel = deque([np.zeros(3, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)
        self.h_grav    = deque([np.zeros(3, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)
        self.h_cmd     = deque([np.zeros(3, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)
        self.h_jpos    = deque([np.zeros(NUM_DOFS, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)
        self.h_jvel    = deque([np.zeros(NUM_DOFS, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)
        self.h_act     = deque([np.zeros(NUM_DOFS, np.float32)] * HISTORY_LEN, maxlen=HISTORY_LEN)

        # IMU gyro
        try:
            mujoco.mj_name2id(self.model, mujoco.mjtObj.mjOBJ_SENSOR, "imu_ang_vel")
            self._has_gyro = True
        except Exception:
            self._has_gyro = False

        self.reset()
        self._setup_viewer()

    # -------------------------------------------------------------------------
    def reset(self):
        mujoco.mj_resetData(self.model, self.data)
        # base flotante a 0.74 con quat identidad
        self.data.qpos[0:3] = [0.0, 0.0, 0.74]
        self.data.qpos[3:7] = [1.0, 0.0, 0.0, 0.0]
        self.data.qpos[7:7 + NUM_DOFS] = DEFAULT_MJ
        self.data.qvel[:] = 0.0
        mujoco.mj_forward(self.model, self.data)
        self.last_action_isaac[:] = 0.0
        for dq in (self.h_ang_vel, self.h_grav, self.h_cmd, self.h_jpos, self.h_jvel, self.h_act):
            for i in range(HISTORY_LEN):
                dq[i][:] = 0.0

    # -------------------------------------------------------------------------
    def _setup_viewer(self):
        self.should_exit = False

        def key_cb(keycode):
            if keycode == 256:  # ESC
                self.should_exit = True
            elif keycode == 265:  # arriba
                self.command[0] += 0.1
            elif keycode == 264:  # abajo
                self.command[0] -= 0.1
            elif keycode == 263:  # izquierda
                self.command[2] += 0.1
            elif keycode == 262:  # derecha
                self.command[2] -= 0.1
            elif keycode in (ord('q'), ord('Q')):
                self.command[1] += 0.1
            elif keycode in (ord('e'), ord('E')):
                self.command[1] -= 0.1
            elif keycode == 32:   # ESPACIO
                self.command[:] = 0.0
            elif keycode in (ord('r'), ord('R')):
                self.reset()
            self.command[:] = np.clip(self.command, -1.0, 1.0)
            print(f"\rComando  vx={self.command[0]:+.2f}  vy={self.command[1]:+.2f}  wz={self.command[2]:+.2f}   ",
                  end="", flush=True)

        self.viewer = mujoco.viewer.launch_passive(
            self.model, self.data, key_callback=key_cb
        )
        self.viewer.cam.distance = 3.0
        self.viewer.cam.elevation = -20
        self.viewer.cam.azimuth = 180

    # -------------------------------------------------------------------------
    def _read_state(self):
        # joints en ORDEN MUJOCO
        qpos_mj = self.data.qpos[7:7 + NUM_DOFS].astype(np.float32)
        qvel_mj = self.data.qvel[6:6 + NUM_DOFS].astype(np.float32)
        quat = self.data.qpos[3:7].astype(np.float32)  # (w,x,y,z)

        if self._has_gyro:
            ang_vel = self.data.sensor("imu_ang_vel").data.astype(np.float32)
        else:
            ang_vel = self.data.qvel[3:6].astype(np.float32)

        return qpos_mj, qvel_mj, quat, ang_vel

    # -------------------------------------------------------------------------
    def _build_obs(self):
        qpos_mj, qvel_mj, quat, ang_vel = self._read_state()

        # --- snapshots del paso actual, ya en ORDEN ISAAC y escalados ---
        ang_vel_obs = ang_vel * SCALE_ANG_VEL                       # (3,)
        grav_obs = quat_rotate_inverse(quat, np.array([0, 0, -1], np.float32))  # (3,)
        cmd_obs = self.command.copy()                              # (3,)

        # joint_pos_rel = (q - default) en orden Isaac
        jpos_isaac = qpos_mj[ISAAC_FROM_MUJOCO] - DEFAULT_ISAAC
        # joint_vel_rel = qd en orden Isaac, escalado
        jvel_isaac = (qvel_mj[ISAAC_FROM_MUJOCO]) * SCALE_JOINT_VEL
        # last_action (orden Isaac, sin escala)
        act_obs = self.last_action_isaac.copy()

        # --- meter al history ---
        self.h_ang_vel.append(ang_vel_obs.astype(np.float32))
        self.h_grav.append(grav_obs.astype(np.float32))
        self.h_cmd.append(cmd_obs.astype(np.float32))
        self.h_jpos.append(jpos_isaac.astype(np.float32))
        self.h_jvel.append(jvel_isaac.astype(np.float32))
        self.h_act.append(act_obs.astype(np.float32))

        # --- aplanar cada termino [t-4, ..., t] y concatenar en el orden del grupo ---
        def flat(dq):
            return np.concatenate(list(dq))  # 5 bloques viejo->nuevo

        obs = np.concatenate([
            flat(self.h_ang_vel),   # 15
            flat(self.h_grav),      # 15
            flat(self.h_cmd),       # 15
            flat(self.h_jpos),      # 120
            flat(self.h_jvel),      # 120
            flat(self.h_act),       # 120
        ]).astype(np.float32)       # total 405

        return obs

    # -------------------------------------------------------------------------
    def _policy_step(self):
        obs = self._build_obs()
        obs_t = torch.from_numpy(obs).unsqueeze(0).to(self.device)
        with torch.no_grad():
            action_isaac = self.policy(obs_t).cpu().numpy().squeeze().astype(np.float32)

        # guardar para la obs del proximo paso (cruda, orden Isaac)
        self.last_action_isaac = action_isaac.copy()

        # target en orden Isaac -> orden MuJoCo
        target_isaac = DEFAULT_ISAAC + ACTION_SCALE * action_isaac
        target_mj = target_isaac[MUJOCO_FROM_ISAAC]
        return target_mj

    # -------------------------------------------------------------------------
    def _apply_pd(self, target_mj):
        qpos_mj = self.data.qpos[7:7 + NUM_DOFS]
        qvel_mj = self.data.qvel[6:6 + NUM_DOFS]
        torque = STIFFNESS_MJ * (target_mj - qpos_mj) - DAMPING_MJ * qvel_mj
        torque = np.clip(torque, -TORQUE_LIMITS_MJ, TORQUE_LIMITS_MJ)
        self.data.ctrl[:NUM_DOFS] = torque

    # -------------------------------------------------------------------------
    def run(self):
        print("\n== R1 con politica NATIVA de Isaac ==")
        print("Flechas: vx/giro | Q/E: lateral | ESPACIO: parar | R: reset | ESC: salir\n")
        target_mj = DEFAULT_MJ.copy()
        step = 0
        try:
            while self.viewer.is_running() and not self.should_exit:
                t0 = time.time()

                if step % DECIMATION == 0:
                    target_mj = self._policy_step()

                self._apply_pd(target_mj)
                mujoco.mj_step(self.model, self.data)

                self.viewer.cam.lookat[:] = self.data.qpos[:3]
                self.viewer.sync()

                step += 1
                dt = SIM_DT - (time.time() - t0)
                if dt > 0:
                    time.sleep(dt)
        except KeyboardInterrupt:
            print("\nInterrumpido")
        finally:
            try:
                if self.viewer.is_running():
                    self.viewer.close()
            except Exception:
                pass
            print("\nFin")


if __name__ == "__main__":
    device = "cpu"
    print(f"Dispositivo: {device}")
    env = R1IsaacPolicyEnv(device=device)
    env.run()